src/JSON/Deserializer.dfy

```diff
@@ -39,68 +39,91 @@ module {:options "-functionSyntax:4"} JSON.Deserializer {
     decreases |str| - start
   { // Assumes UTF-16 strings
-    if start >= |str| then Success([])
-    else if str[start] == '\\' then
+    if start >= |str| then Success(prefix)
+    else if str[start] == '\\' as uint16 then
       if |str| == start + 1 then
         Failure(EscapeAtEOS)
       else
         var c := str[start + 1];
-        if c == 'u' then
+        if c == 'u' as uint16 then
           if |str| <= start + 6 then
             Failure(EscapeAtEOS)
           else
             var code := str[start + 2..start + 6];
-            if exists c | c in code :: c !in Str.HEX_TABLE then
-              Failure(UnsupportedEscape(code))
+            if exists c | c in code :: c !in HEX_TABLE_16 then
+              Failure(UnsupportedEscape16(code))
             else
-              var tl :- Unescape(str, start + 6);
-              var hd := Str.ToNat(code, 16);
-              assert hd < 0x10000 by { reveal Pow(); }
-              Success([hd as char] + tl)
+              var hd := ToNat16(code);
+              Unescape(str, start + 6, prefix + [hd])
         else
           var unescaped: uint16 := match c
-             case '\"' => 0x22 as uint16 // quotation mark
-             case '\\' => 0x5C as uint16 // reverse solidus
-             case 'b'  => 0x08 as uint16 // backspace
-             case 'f'  => 0x0C as uint16 // form feed
-             case 'n'  => 0x0A as uint16 // line feed
-             case 'r'  => 0x0D as uint16 // carriage return
-             case 't'  => 0x09 as uint16 // tab
-             case _    => 0 as uint16;
-          if unescaped == 0 as uint16 then
-            Failure(UnsupportedEscape(str[start..start+2]))
+            case 0x22 => 0x22 as uint16 // \" => quotation mark
+            case 0x5C => 0x5C as uint16 // \\ => reverse solidus
+            case 0x62 => 0x08 as uint16 // \b => backspace
+            case 0x66 => 0x0C as uint16 // \f => form feed
+            case 0x6E => 0x0A as uint16 // \n => line feed
+            case 0x72 => 0x0D as uint16 // \r => carriage return
+            case 0x74 => 0x09 as uint16 // \t => tab
+            case _    => 0 as uint16;
+          if unescaped as int == 0 then
+            Failure(UnsupportedEscape16(str[start..start+2]))
           else
-            var tl :- Unescape(str, start + 2);
-            Success([unescaped as char] + tl)
+            Unescape(str, start + 2, prefix + [unescaped])
     else
-      var tl :- Unescape(str, start + 1);
-      Success([str[start]] + tl)
-  }
-
-  function Transcode8To16Unescaped(str: seq<byte>): DeserializationResult<string>
-    // TODO Optimize with a function by method
-  {
-    Unescape(Unicode.Transcode8To16(str))
+      Unescape(str, start + 1, prefix + [str[start]])
   }
 
   function String(js: Grammar.jstring): DeserializationResult<string> {
-    Transcode8To16Unescaped(js.contents.Bytes())
+    // TODO Optimize with a function by method
```

Comment:
I removed it: since that TODO was left there, I make Unescape tail-recursive to address the worst performance issue. There's definitely more to do here and in the Unicode module, but that's a more widespread challenge to take on in the future.